src/voice/input/cached/mod.rs

```diff
@@ -0,0 +1,1533 @@
+/// retrieved in **uncompressed floating point** form to minimise the
+/// cost of audio processing. This is a significant *3 Mbps (375 kiB/s)*,
+/// or 131 MiB of RAM for a 6 minute song.
+///
+/// [`Input`]: ../struct.Input.html
+/// [`CompressedSource`]: struct.CompressedSource.html
+/// [`RestartableSource`]: ../struct.RestartableSource.html
+#[derive(Clone, Debug)]
+pub struct MemorySource {
+    cache: AudioCache,
+}
+
+impl MemorySource {
+    /// Wrap an existing [`Input`] with an in-memory store of raw 32-bit floating point audio.
+    ///
+    /// `length_hint` may be used to control the size of the initial chunk, preventing
+    /// needless allocations and copies. If this is not present, the value specified in
+    /// `source`'s [`Metadata.duration`] will be used.
+    ///
+    /// [`Input`]: ../struct.Input.html
+    /// [`Metadata.duration`]: ../struct.Metadata.html#structfield.duration
+    pub fn new(source: Input, config: Option<CacheConfig>) -> Self {
+        let core_raw = RawStore::new(source, EncodingData::FloatPcm, config)
+            .expect("This should only be fallible for Opus caches.");
+
+        Self {
+            cache: AudioCache::new(core_raw),
+        }
+    }
+
+    /// Acquire a new handle to this object, to begin a new
+    /// source from the exsting cached data.
+    pub fn new_handle(&self) -> Self {
+        Self {
+            cache: self.cache.new_handle(),
+        }
+    }
+
+    /// Block the current thread to read all bytes from the underlying stream
+    /// into the backing store.
+    pub fn load_file(&mut self) {
+        self.cache.load_file();
+    }
+
+    /// Spawn a new thread to read all bytes from the underlying stream
+    /// into the backing store.
+    pub fn spawn_loader(&self) -> std::thread::JoinHandle<()> {
+        self.cache.spawn_loader()
+    }
+}
+
+impl From<MemorySource> for Input {
+    fn from(src: MemorySource) -> Self {
+        let metadata = src.cache.core.metadata()
+            .map(|m| m.clone())
+            .unwrap_or_default();
+
+        Self {
+            metadata,
+            stereo: src.cache.core.is_stereo(),
+            kind: Codec::FloatPcm,
+            container: Container::Raw,
+
+            reader: Reader::InMemory(src),
+        }
+    }
+}
+
+impl Read for MemorySource {
+    fn read(&mut self, buf: &mut [u8]) -> IoResult<usize> {
+        self.cache.read(buf)
+    }
+}
+
+impl Seek for MemorySource {
+    fn seek(&mut self, pos: SeekFrom) -> IoResult<u64> {
+        let old_pos = self.cache.pos as u64;
+
+        let (valid, new_pos) = match pos {
+            SeekFrom::Current(adj) => {
+                // overflow expected in many cases.
+                let new_pos = old_pos.wrapping_add(adj as u64);
+                (adj >= 0 || (adj.abs() as u64) <= old_pos, new_pos)
+            }
+            SeekFrom::End(adj) => {
+                // FIXME: make this check for metadata as the basis?
```

Comment:
Any further plans for this?

Reply:
I think it might be better to do this in the slow (and correct) way we have right now, where the file is loaded to the end first. It will be tricky to defensively program around possibly mis-reported durations etc. Time-based seeking always uses `SeekFrom::Start(..)`, so the other variants are mainly implemented for the sake of completeness.